File: backend/src/main/java/ma/zs/alc/service/facade/collaborator/collab/CollaboratorCollaboratorService.java
package ma.zs.alc.service.facade.collaborator.collab;

import java.util.List;
import ma.zs.alc.bean.core.collab.Collaborator;
import ma.zs.alc.dao.criteria.core.collab.CollaboratorCriteria;
import ma.zs.alc.zynerator.dto.AccountValidationDto;
import ma.zs.alc.zynerator.security.payload.response.JwtResponse;
import ma.zs.alc.zynerator.service.IService;
import org.springframework.http.ResponseEntity;


public interface CollaboratorCollaboratorService extends  IService<Collaborator,CollaboratorCriteria>  {
    Collaborator register(Collaborator t);

    Collaborator findByUsername(String email);
    boolean changePassword(String username, String newPassword);

    List<Collaborator> findByTypeCollaboratorId(Long id);
    int deleteByTypeCollaboratorId(Long id);
    long countByTypeCollaboratorCode(String code);

    boolean updateAccountLockStatus(Long id, boolean accountNonLocked);
    boolean updateAccountNonExpiredStatus(Long id, boolean accountNonExpired);
    boolean updateAccountStatus(Long id, boolean enabled);
    boolean updatePasswordChangedStatus(Long id, boolean passwordChanged);
    boolean updateCredentialsNonExpiredStatus(Long id, boolean credentialsNonExpired);
    boolean onCommunicationEnabled(Long id, boolean communicationEnabled);
    boolean onSecurityEnabled(Long id, boolean securityEnabled);
    boolean onLessonReminderEnabled(Long id, boolean lessonReminderEnabled);
    boolean onClassroomEnabled(Long id, boolean classroomEnabled);
    boolean onPasswordChangedNotificationEnabled(Long id, boolean passwordChangedNotificationEnabled);
    boolean onContactNotificationEnabled(Long id, boolean contactNotificationEnabled);

    boolean findByUsernameAndValidationCode(String username, String validationCode);

    ResponseEntity<JwtResponse> validateUser(AccountValidationDto accountValidationDto);
}

File: frontend/src/app/shared/service/collaborator/vocab/CollaboratorCollaborator.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';

import {environment} from '../../../../../environments/environment';

import {CollaboratorDto} from '../../../model/vocab/Collaborator.model';
import {CollaboratorCriteria} from '../../../criteria/vocab/CollaboratorCriteria.model';
import {AbstractService} from "../../../../zynerator/service/AbstractService";
import {catchError, Observable, ReplaySubject, tap, throwError} from "rxjs";
import {ProfDto} from "../../../model/prof/Prof.model";
import {EtudiantDto} from "../../../model/inscription/Etudiant.model";


@Injectable({
    providedIn: 'root'
})
export class CollaboratorCollaboratorService extends AbstractService<CollaboratorDto, CollaboratorCriteria> {

    private _user: ReplaySubject<CollaboratorDto> = new ReplaySubject<CollaboratorDto>(1);
    constructor(private http: HttpClient) {
        super();
        this.setHttp(http);
    }

    get API() {
        return environment.apiUrlAlcservice + 'collaborator/collaborator/';
    }

    public constrcutDto(): CollaboratorDto {
        return new CollaboratorDto();
    }

    public constrcutCriteria(): CollaboratorCriteria {
        return new CollaboratorCriteria();
    }

    get(id: number): Observable<CollaboratorDto> {
        return this.http.get<CollaboratorDto>(`${this.API}id/${id}`).pipe(
            tap((collaborator) => {
                this._user.next(collaborator);
            }),
        );
    }
    updatePasswordChangedStatus(id: number, passwordChanged: boolean): Observable<any> {
        return this.http.patch(`${this.API}update-password-changed/${id}`, { passwordChanged });
    }
    updateAccountNonExpiredStatus(id: number, accountNonExpired: boolean): Observable<any> {
        return this.http.patch(`${this.API}update-account-non-expired/${id}`, { accountNonExpired });
    }

    updateAccountStatus(id: number, enabled: boolean): Observable<any> {
        return this.http.patch(`${this.API}update-status/${id}`, { enabled });
    }

    updateCredentialsNonExpiredStatus(id: number, credentialsNonExpired: boolean): Observable<any> {
        return this.http.patch(`${this.API}update-credentials-non-expired/${id}`, { credentialsNonExpired });

    }
    changePassword(username: string, newPassword: string): Observable<any> {
        return this.http.put(`${this.API}change-password`, { username, newPassword });
    }
    updateAccountLockStatus(userId: number, accountNonLocked: boolean): Observable<any> {
        return this.http.patch(`${this.API}update-lock-status/${userId}`, { accountNonLocked });
    }

    findByUserName(email: string): Observable<CollaboratorDto> {
        return this.http.get<CollaboratorDto>(`${this.API}username/${email}`).pipe(
            catchError(error => {
                console.error('Error fetching collab:', error);
                return throwError(error);
            })
        );
    }
    onCommunicationEnabled(userId: number, communicationEnabled: boolean): Observable<any> {
        return this.http.patch(`${this.API}communicationEnabled/${userId}`, { communicationEnabled });
    }
    onSecurityEnabled(userId: number, securityEnabled: boolean): Observable<any> {
        return this.http.patch(`${this.API}securityEnabled/${userId}`, { securityEnabled });
    }
    onLessonReminderEnabled(userId: number, lessonReminderEnabled: boolean): Observable<any> {
        return this.http.patch(`${this.API}lessonReminderEnabled/${userId}`, { lessonReminderEnabled });
    }
    onClassroomEnabled(userId: number, classroomEnabled: boolean): Observable<any> {
        return this.http.patch(`${this.API}classroomEnabled/${userId}`, { classroomEnabled });
    }
    onPasswordChangedNotificationEnabled(userId: number, passwordChangedNotificationEnabled: boolean): Observable<any> {
        return this.http.patch(`${this.API}passwordChangedNotificationEnabled/${userId}`, { passwordChangedNotificationEnabled });
    }
    onContactNotificationEnabled(userId: number, contactNotificationEnabled: boolean): Observable<any> {
        return this.http.patch(`${this.API}contactNotificationEnabled/${userId}`, { contactNotificationEnabled });
    }
}
